File: trifecta-android/app/src/main/java/com/belweave/trifecta/core/networking/EffectRpc.kt
package com.belweave.trifecta.core.networking

import com.belweave.trifecta.core.models.T3Json
import com.belweave.trifecta.core.models.asArrayOrNull
import com.belweave.trifecta.core.models.asObjectOrNull
import com.belweave.trifecta.core.models.bool
import com.belweave.trifecta.core.models.intAt
import com.belweave.trifecta.core.models.obj
import com.belweave.trifecta.core.models.stringOrNull
import com.belweave.trifecta.core.models.str
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.util.UUID

/**
 * Effect-style RPC wire protocol used by the Trifecta desktop server.
 * Mirrors the iOS `EffectRPCMessage` shape so a single message gets
 * encoded as a single-message JSON object (not an array).
 */
sealed class EffectRpcMessage {
    abstract val isOutbound: Boolean

    data class Request(
        val id: String,
        val tag: String,
        val payload: JsonElement,
        val headers: List<List<String>> = emptyList()
    ) : EffectRpcMessage() {
        override val isOutbound: Boolean = true
    }

    data class StreamRequest(
        val id: String,
        val tag: String,
        val payload: JsonElement,
        val headers: List<List<String>> = emptyList()
    ) : EffectRpcMessage() {
        override val isOutbound: Boolean = true
    }

    data class Interrupt(val requestId: String, val interruptors: List<String>) : EffectRpcMessage() {
        override val isOutbound: Boolean = true
    }

    data class Ack(val requestId: String) : EffectRpcMessage() {
        override val isOutbound: Boolean = true
    }

    object Ping : EffectRpcMessage() { override val isOutbound: Boolean = true }
    object Pong : EffectRpcMessage() { override val isOutbound: Boolean = true }
    object Eof : EffectRpcMessage() { override val isOutbound: Boolean = true }

    data class Chunk(val requestId: String, val values: List<JsonElement>) : EffectRpcMessage() {
        override val isOutbound: Boolean = false
    }

    data class Exit(
        val requestId: String,
        val success: Boolean,
        val value: JsonElement?,
        val errorTag: String?,
        val errorMessage: String?
    ) : EffectRpcMessage() {
        override val isOutbound: Boolean = false
    }

    data class Defect(val message: String) : EffectRpcMessage() {
        override val isOutbound: Boolean = false
    }

    data class Unknown(val raw: JsonElement) : EffectRpcMessage() {
        override val isOutbound: Boolean = false
    }
}

object EffectRpcEncoder {
    fun encode(message: EffectRpcMessage): String =
        T3Json.encodeToString(JsonElement.serializer(), encodeOne(message))

    private fun encodeOne(msg: EffectRpcMessage): JsonElement = when (msg) {
        is EffectRpcMessage.Request -> buildJsonObject {
            put("_tag", "Request")
            put("id", msg.id)
            put("tag", msg.tag)
            put("payload", msg.payload)
            put("headers", buildJsonArray {
                msg.headers.forEach { entry ->
                    add(buildJsonArray { entry.forEach { add(JsonPrimitive(it)) } })
                }
            })
            put("spanId", randomHex(16))
            put("traceId", randomHex(32))
            put("sampled", false)
        }
        is EffectRpcMessage.StreamRequest -> buildJsonObject {
            put("_tag", "Request")
            put("id", msg.id)
            put("tag", msg.tag)
            put("payload", msg.payload)
            put("headers", buildJsonArray {
                msg.headers.forEach { entry ->
                    add(buildJsonArray { entry.forEach { add(JsonPrimitive(it)) } })
                }
            })
            put("spanId", randomHex(16))
            put("traceId", randomHex(32))
            put("sampled", false)
        }
        is EffectRpcMessage.Interrupt -> buildJsonObject {
            put("_tag", "Interrupt")
            put("requestId", msg.requestId)
            put("interruptors", buildJsonArray { msg.interruptors.forEach { add(JsonPrimitive(it)) } })
        }
        is EffectRpcMessage.Ack -> buildJsonObject {
            put("_tag", "Ack")
            put("requestId", msg.requestId)
        }
        EffectRpcMessage.Ping -> buildJsonObject { put("_tag", "Ping") }
        EffectRpcMessage.Pong -> buildJsonObject { put("_tag", "Pong") }
        EffectRpcMessage.Eof -> buildJsonObject { put("_tag", "Eof") }
        else -> buildJsonObject { }
    }

    private fun randomHex(count: Int): String =
        UUID.randomUUID().toString().replace("-", "").lowercase().take(count)
}

object EffectRpcDecoder {
    fun decodeFrame(text: String): List<EffectRpcMessage> {
        val element = T3Json.parseToJsonElement(text)
        val array = element.asArrayOrNull()
        return if (array != null) array.map(::decodeOne) else listOf(decodeOne(element))
    }

    private fun decodeOne(element: JsonElement): EffectRpcMessage {
        val obj = element.asObjectOrNull() ?: return EffectRpcMessage.Unknown(element)
        return when (obj.str("_tag")) {
            "Ping" -> EffectRpcMessage.Ping
            "Pong" -> EffectRpcMessage.Pong
            "Chunk" -> {
                val id = obj.str("requestId") ?: ""
                val values = obj["values"]?.asArrayOrNull() ?: JsonArray(emptyList())
                EffectRpcMessage.Chunk(id, values.toList())
            }
            "Exit" -> {
                val id = obj.str("requestId") ?: ""
                val exit = obj.obj("exit") ?: JsonObject(emptyMap())
                val exitTag = exit.str("_tag") ?: "Failure"
                if (exitTag == "Success") {
                    EffectRpcMessage.Exit(
                        requestId = id,
                        success = true,
                        value = exit["value"],
                        errorTag = null,
                        errorMessage = null
                    )
                } else {
                    val cause = exit.obj("cause") ?: JsonObject(emptyMap())
                    val error = findErrorPayload(cause)
                    val errorTag = error?.str("_tag")
                    val errorMessage = taggedErrorMessage(errorTag, error)
                        ?: findErrorMessage(cause)
                        ?: "Server error"
                    EffectRpcMessage.Exit(
                        requestId = id,
                        success = false,
                        value = null,
                        errorTag = errorTag,
                        errorMessage = errorMessage
                    )
                }
            }
            "Defect" -> EffectRpcMessage.Defect(obj.str("defect") ?: "Unknown server defect")
            "ClientProtocolError" -> EffectRpcMessage.Defect(obj.str("error") ?: "Client protocol error")
            else -> EffectRpcMessage.Unknown(element)
        }
    }

    private fun findErrorPayload(value: JsonElement?): JsonObject? {
        val obj = value.asObjectOrNull()
        if (obj != null) {
            obj.obj("error")?.let { return it }
            obj.values.forEach { child ->
                val found = findErrorPayload(child)
                if (found != null) return found
            }
            return null
        }
        val arr = value.asArrayOrNull() ?: return null
        arr.forEach { child ->
            val found = findErrorPayload(child)
            if (found != null) return found
        }
        return null
    }

    private fun findErrorMessage(value: JsonElement?): String? {
        val obj = value.asObjectOrNull()
        if (obj != null) {
            val directKeys = listOf("message", "detail", "reason", "description", "errorDescription", "defect")
            directKeys.forEach { key ->
                val message = obj.str(key)
                if (!message.isNullOrBlank() && !isRawFingerprint(message)) return message
            }
            listOf("error", "cause", "failure", "defect").forEach { key ->
                findErrorMessage(obj[key])?.let { return it }
            }
            obj.values.forEach { child ->
                findErrorMessage(child)?.let { return it }
            }
            return null
        }
        val arr = value.asArrayOrNull()
        if (arr != null) {
            arr.forEach { child ->
                findErrorMessage(child)?.let { return it }
            }
            return null
        }
        val raw = value?.stringOrNull()
        return if (!raw.isNullOrBlank() && !isRawFingerprint(raw)) raw else null
    }

    private fun taggedErrorMessage(tag: String?, error: JsonObject?): String? = when (tag) {
        "SshHostProfileNotFoundError" ->
            "SSH host profile was not found. Refresh the host list and try again."
        "SshHostProfileConflictError" ->
            "An SSH host with this label or address already exists."
        "SshHostKeyMismatchError" -> {
            val hostname = error?.str("hostname") ?: "host"
            val port = error?.intAt("port") ?: 22
            "Host key mismatch for $hostname:$port. Refusing to connect."
        }
        "SshSessionNotFoundError" ->
            "SSH session is no longer active. Reconnect and try again."
        "SshSessionTokenInvalidError" -> {
            val reason = error?.str("reason") ?: "invalid"
            "SSH session token rejected ($reason). Reconnect and try again."
        }
        "SshAuthorizationError" -> {
            val reason = error?.str("reason") ?: "not authorized"
            "SSH operation denied: $reason"
        }
        "SshSessionLimitError" -> {
            val limit = error?.intAt("limit") ?: 0
            if (limit > 0) "SSH session limit reached ($limit). Close another SSH session and try again."
            else "SSH session limit reached. Close another SSH session and try again."
        }
        else -> null
    }

    private fun isRawFingerprint(value: String): Boolean =
        value.startsWith("SHA256:") && !value.contains(" ")
}

File: trifecta-ios/Trifecta/Core/Networking/EffectRPC.swift
import Foundation

enum EffectRPCMessage {
    case request(id: String, tag: String, payload: Any, headers: [[String]])
    case streamRequest(id: String, tag: String, payload: Any, headers: [[String]])
    case interrupt(requestId: String, interruptors: [String])
    case ack(requestId: String)
    case ping
    case eof

    case chunk(requestId: String, values: [Any])
    case exit(requestId: String, success: Bool, value: Any?, errorTag: String?, errorMessage: String?)
    case pong
    case defect(message: String)
    case unknown(json: Any)

    var isOutbound: Bool {
        switch self {
        case .request, .streamRequest, .interrupt, .ack, .ping, .pong, .eof: true
        default: false
        }
    }
}

enum EffectRPCEncoder {
    static func encode(_ message: EffectRPCMessage) throws -> Data {
        try JSONSerialization.data(withJSONObject: encodeOne(message), options: [])
    }

    private static func encodeOne(_ msg: EffectRPCMessage) -> [String: Any] {
        switch msg {
        case let .request(id, tag, payload, headers):
            return ["_tag": "Request", "id": id, "tag": tag,
                    "payload": payload, "headers": headers,
                    "spanId": randomHex(count: 16), "traceId": randomHex(count: 32),
                    "sampled": false]
        case let .streamRequest(id, tag, payload, headers):
            return ["_tag": "Request", "id": id, "tag": tag,
                    "payload": payload, "headers": headers,
                    "spanId": randomHex(count: 16), "traceId": randomHex(count: 32),
                    "sampled": false]
        case let .interrupt(requestId, interruptors):
            return ["_tag": "Interrupt", "requestId": requestId, "interruptors": interruptors]
        case let .ack(requestId):
            return ["_tag": "Ack", "requestId": requestId]
        case .ping:
            return ["_tag": "Ping"]
        case .pong:
            return ["_tag": "Pong"]
        case .eof:
            return ["_tag": "Eof"]
        default:
            return [:]
        }
    }

    private static func randomHex(count: Int) -> String {
        UUID().uuidString
            .replacingOccurrences(of: "-", with: "")
            .lowercased()
            .prefix(count)
            .description
    }
}

enum EffectRPCDecoder {
    static func decodeFrame(_ data: Data) throws -> [EffectRPCMessage] {
        let raw = try JSONSerialization.jsonObject(with: data)
        if let array = raw as? [Any] {
            return try array.map(decodeOne)
        }
        return [try decodeOne(raw)]
    }

    private static func decodeOne(_ any: Any) throws -> EffectRPCMessage {
        guard let dict = any as? [String: Any], let tag = dict["_tag"] as? String else {
            return .unknown(json: any)
        }
        switch tag {
        case "Ping":
            return .ping
        case "Pong":
            return .pong
        case "Chunk":
            let id = (dict["requestId"] as? String) ?? ""
            let values = (dict["values"] as? [Any]) ?? []
            return .chunk(requestId: id, values: values)
        case "Exit":
            let id = (dict["requestId"] as? String) ?? ""
            let exit = dict["exit"] as? [String: Any] ?? [:]
            let exitTag = exit["_tag"] as? String ?? "Failure"
            if exitTag == "Success" {
                return .exit(requestId: id, success: true, value: exit["value"],
                             errorTag: nil, errorMessage: nil)
            } else {
                let cause = exit["cause"] as? [String: Any] ?? [:]
                let error = findErrorPayload(in: cause)
                let errorMessage = findErrorMessage(in: cause) ?? "Server error"
                let errorTag = error?["_tag"] as? String
                return .exit(requestId: id, success: false, value: nil,
                             errorTag: errorTag, errorMessage: errorMessage)
            }
        case "Defect":
            let message = (dict["defect"] as? String) ?? "Unknown server defect"
            return .defect(message: message)
        case "ClientProtocolError":
            let message = (dict["error"] as? String) ?? "Client protocol error"
            return .defect(message: message)
        default:
            return .unknown(json: any)
        }
    }

    private static func findErrorPayload(in value: Any) -> [String: Any]? {
        if let dict = value as? [String: Any] {
            if let error = dict["error"] as? [String: Any] {
                return error
            }
            for child in dict.values {
                if let found = findErrorPayload(in: child) {
                    return found
                }
            }
        } else if let array = value as? [Any] {
            for child in array {
                if let found = findErrorPayload(in: child) {
                    return found
                }
            }
        }
        return nil
    }

    private static func findErrorMessage(in value: Any) -> String? {
        if let dict = value as? [String: Any] {
            if let message = taggedErrorMessage(in: dict) {
                return message
            }
            for key in ["message", "detail", "reason", "description", "errorDescription", "defect"] {
                if let message = dict[key] as? String,
                   !message.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
                   !isRawFingerprint(message) {
                    return message
                }
            }
            for key in ["error", "cause", "failure", "defect"] {
                if let child = dict[key],
                   let found = findErrorMessage(in: child) {
                    return found
                }
            }
            for (key, child) in dict where !ignoredErrorValueKeys.contains(key) {
                if let found = findErrorMessage(in: child) {
                    return found
                }
            }
        } else if let array = value as? [Any] {
            for child in array {
                if let found = findErrorMessage(in: child) {
                    return found
                }
            }
        } else if let message = value as? String,
                  !message.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
                  !isRawFingerprint(message) {
            return message
        }
        return nil
    }

    private static let ignoredErrorValueKeys = Set([
        "_tag", "id", "requestId", "tag", "sessionId", "hostId",
        "fingerprintSha256", "expectedFingerprint", "actualFingerprint",
    ])

    private static func taggedErrorMessage(in dict: [String: Any]) -> String? {
        guard let tag = dict["_tag"] as? String else { return nil }
        switch tag {
        case "SshHostProfileNotFoundError":
            return "SSH host profile was not found. Refresh the host list and try again."
        case "SshHostProfileConflictError":
            return "An SSH host with this label or address already exists."
        case "SshHostKeyMismatchError":
            let hostname = dict["hostname"] as? String ?? "host"
            let port = dict["port"].map { "\($0)" } ?? "22"
            return "Host key mismatch for \(hostname):\(port). Refusing to connect."
        case "SshSessionNotFoundError":
            return "SSH session is no longer active. Reconnect and try again."
        case "SshSessionTokenInvalidError":
            let reason = dict["reason"] as? String ?? "invalid"
            return "SSH session token rejected (\(reason)). Reconnect and try again."
        case "SshAuthorizationError":
            let reason = dict["reason"] as? String ?? "not authorized"
            return "SSH operation denied: \(reason)"
        case "SshSessionLimitError":
            let limit = dict["limit"].map { "\($0)" } ?? "maximum"
            return "SSH session limit reached (\(limit)). Close another SSH session and try again."
        default:
            return nil
        }
    }

    private static func isRawFingerprint(_ text: String) -> Bool {
        text.hasPrefix("SHA256:") && !text.contains(" ")
    }
}
